### PR TITLE
Adding support for Linux s390x

### PR DESCRIPTION
**Pull request by namrata-ibm**
Adding support for Linux s390x for TensorFlow

### CODE REVIEW

**Comment by gunan** [comment]
Thank you for recreating the PR.
CLA looks good.
Jenkins, test this please.

**Comment by gunan** [comment]
Windows cmake test failure might be an issue on jenkins executor.
Jenkins, test this please.

**Comment by gunan** [comment]
https://ci.tensorflow.org/job/tensorflow-pr-win-cmake-py/704/console

This time it failed for real. Could you take a look?

**Comment by namrata-ibm** [comment]
The error is '__BYTE_ORDER__' and '__ORDER_BIG_ENDIAN__' are undeclared identifiers. However @guschmue had already committed the fix for BYTE_ORDER on windows through [PR](https://github.com/tensorflow/tensorflow/pull/6102). 
@gushmue could you please have a look?

**Comment by namrata-ibm** [comment]
Do i need to explicitly include `tensorflow/core/platform/cpu_info.h` in the files where I am using the BYTE_ORDER flag?

**Comment by gunan** [comment]
To me it looks like you definitely need to include `tensorflow/core/platform/cpu_info.h`. You also need to add a BUILD dependency for the library which exposes this header file.

File: tensorflow/core/framework/bfloat16.cc [hunk 1]
```diff
@@ -20,17 +20,30 @@ namespace tensorflow {
 void FloatToBFloat16(const float* src, bfloat16* dst, int64 size) {
   const uint16_t* p = reinterpret_cast<const uint16_t*>(src);
   uint16_t* q = reinterpret_cast<uint16_t*>(dst);
-  for (; size; p += 2, q++, size--) {
-    *q = p[1];
+  if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {
```
Comment (drpngx) [inline tensorflow/core/framework/bfloat16.cc]: Here and everywhere: we should probably use #if.

File: tensorflow/core/framework/bfloat16.cc [hunk 2]
```diff
@@ -20,17 +20,30 @@ namespace tensorflow {
...
   const uint16_t* p = reinterpret_cast<const uint16_t*>(src);
   uint16_t* q = reinterpret_cast<uint16_t*>(dst);
-  for (; size; p += 2, q++, size--) {
-    *q = p[1];
+  if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {  
+    for (; size; p += 2, q++, size--) {  
+      *q = p[0];  
+    }  
+  } else {  
+    for (; size; p += 2, q++, size--) {
```
Comment (drpngx) [inline tensorflow/core/framework/bfloat16.cc]: I know it was in the original code, but please correct the style:
`size != 0`

File: tensorflow/core/kernels/cast_op.h [hunk 3]
```diff
@@ -96,11 +96,16 @@ struct scalar_cast_op<::tensorflow::bfloat16, float> {
...
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE float operator()(
       const ::tensorflow::bfloat16& a) const {
-    static_assert(::tensorflow::port::kLittleEndian, "");
     float ret;
     uint16_t* p = reinterpret_cast<uint16_t*>(&ret);
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+    p[0] = a.value;  
+    p[1] = 0;  
+#else  
+    static_assert(::tensorflow::port::kLittleEndian, "");
```
Comment (drpngx) [inline tensorflow/core/kernels/cast_op.h]: Sorry, while we're at it, could we add an error message to the assert here?

File: tensorflow/python/framework/tensor_util_test.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 from __future__ import print_function
 
 import numpy as np
+import sys
```
Comment (drpngx) [inline tensorflow/python/framework/tensor_util_test.py]: Do you need this?

File: tensorflow/python/kernel_tests/cast_op_test.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
 from __future__ import print_function
 
 import numpy as np
+import sys
```
Comment (drpngx) [inline tensorflow/python/kernel_tests/cast_op_test.py]: I don't think you need this.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/core/framework/bfloat16.cc (hunk 1 above)
Comment (namrata-ibm) [inline tensorflow/core/framework/bfloat16.cc]: ok, we will work on this change.

Comment (namrata-ibm) [inline tensorflow/core/framework/bfloat16.cc]: ok, we will work on this change.

File: tensorflow/python/framework/tensor_util_test.py (hunk 4 above)
Comment (namrata-ibm) [inline tensorflow/python/framework/tensor_util_test.py]: We need to import sys as we are using 'sys.byteorder' to detect endianness of underlying platform.

Comment (namrata-ibm) [inline tensorflow/python/framework/tensor_util_test.py]: We need to import sys as we are using 'sys.byteorder' to detect endianness of underlying platform.

File: tensorflow/python/kernel_tests/cast_op_test.py (hunk 5 above)
Comment (namrata-ibm) [inline tensorflow/python/kernel_tests/cast_op_test.py]: We need to import sys as we are using 'sys.byteorder' to detect endianness of underlying platform.

File: tensorflow/core/framework/bfloat16.cc (hunk 2 above)
Comment (namrata-ibm) [inline tensorflow/core/framework/bfloat16.cc]: ok

File: tensorflow/core/kernels/cast_op.h (hunk 3 above)
Comment (namrata-ibm) [inline tensorflow/core/kernels/cast_op.h]: ok

**Comment by drpngx** [comment]
Sounds good, will merge once the build is restored.

**Comment by drpngx** [comment]
09:39:59          c:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\tensorflow\core\framework\bfloat16.cc(23): error C2065: '__BYTE_ORDER__': undeclared identifier [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\tf_core_framework.vcxproj]
09:39:59          c:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\tensorflow\core\framework\bfloat16.cc(23): error C2065: '__ORDER_BIG_ENDIAN__': undeclared identifier [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\tf_core_framework.vcxproj]
09:39:59          c:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\tensorflow\core\framework\bfloat16.cc(37): error C2065: '__BYTE_ORDER__': undeclared identifier [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\tf_core_framework.vcxproj]
09:39:59          c:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\tensorflow\core\framework\bfloat16.cc(37): error C2065: '__ORDER_BIG_ENDIAN__': undeclared identifier [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\tf_core_framework.vcxproj]

**Comment by namrata-ibm** [comment]
@drpngx, I have made changes as per your feedback. 
@gunan, `cpu_info.h` has been included in all respective files and as a BUILD dependency. 
Could you request a build to verify?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Please put the `cpu_info` alphabetically sorted in `tensorflow/core/BUILD`

**Comment by drpngx** [comment]
Jenkins, test this please.

**Pull request merged**